### DISCUSSION TITLE
Do we need a FastPointFunc and LoosePointFunc?

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pymc3#5237] [General]
It seems that the difference is that `LoosePointFunc` allows you to evaluate a  `logp`/`dlogp`/ what have you with args and kwargs, instead of just passing a dictionary, but at the cost of being slower because it has to wrap them in a dictionary (Point) before calling logp.

```python
import pymc as pm

with pm.Model() as m:
    x = pm.Normal('x')
    y = pm.Normal('y', 0, 1, observed=0)

# Uses `LoosePointFunc` under the hood
m.logp({'x': 0})
m.logp(x=0)
m.logp([('x', 0)])  # Couldn't think of any other way of using `args`

# Uses `FastPointFunc` under the hood
m.fastlogp({'x': 0})
m.fastlogp(x=0) # Fails
m.fastlogp([('x', 0)]) # Fails
```

Does anyone actually use this? Should the default be the "fast" version that only accepts a dictionary, and the flexible one be called something like "slow"? Do we even need the slow version?

https://github.com/pymc-devs/pymc/blob/f6f1a8e807a0719a07fdb17ed79af9adce065814/pymc/model.py#L1722-L1766

**Comment by twiecki** [comment]
Yeah, I think this is quite confusing, if we can remove them that would be ideal.

**Comment by ricardoV94** [comment]
We could start deprecating warnings for args and kwargs in V3 and remove the LoosePointFunc in V4

**Comment by ricardoV94** [comment]
More importantly, these properties/methods always compile a new aesara function which is quite slow if one wants to evaluate a model many times.

We might want to rename them to something like `model.compile_logp` to make this obvious and hint that this shouldn't be repeatedly invoked but cached once and then reused.

Similarly to how we have `recompute_initial_point` that indicates this is a costly operation.

**Comment by twiecki** [comment]
This is done, right?

**Comment by ricardoV94** [comment]
No, but we can open an issue
